src/psyclone/psyir/nodes/structure_member.py

```diff
@@ -0,0 +1,137 @@
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the implementation of the StructureMember node.'''
+
+from __future__ import absolute_import
+from psyclone.psyir.nodes.member import Member
+from psyclone.psyir.symbols import TypeSymbol
+from psyclone.psyir.symbols.datatypes import StructureType, DeferredType, \
+    ArrayType
+
+
+class StructureMember(Member):
+    '''
+    Node representing a membership expression of the parent's Reference that
+    resolves into another structure.
+
+    :param str member: the name of the structure member that is \
+                       being accessed.
+    :param struct_type: the datatype of the parent structure.
+    :type struct_type: :py:class:`psyclone.psyir.symbols.StructureType` or \
+                       :py:class:`psyclone.psyir.symbols.TypeSymbol`
+    :param parent: the parent of this node in the PSyIR tree.
+    :type parent: :py:class:`psyclone.psyir.nodes.StructureReference` or \
+                  :py:class:`psyclone.psyir.nodes.Member`
+
+    :raises TypeError: if the type of the specified member is not \
+                       consistent with it being a structure type.
+
+    '''
+    # Textual description of the node. Since it represents a reference to a
+    # structure it may have a single child which is a reference to one of its
+    # members.
+    _children_valid_format = "[Member]"
+    _text_name = "StructureMember"
+
+    def __init__(self, member, struct_type=None, parent=None):
+
+        super(StructureMember, self).__init__(member,
+                                              struct_type=struct_type,
+                                              parent=parent)
+        if self.component:
+            if isinstance(self.component.datatype, ArrayType):
+                target_type = self.component.datatype.intrinsic
+            else:
+                target_type = self.component.datatype
+                if not isinstance(target_type,
+                                  (DeferredType, StructureType, TypeSymbol)):
+                    raise TypeError(
+                        "The member '{0}' is not of DeferredType, "
+                        "StructureType or a TypeSymbol and therefore cannot "
+                        "be the target of a StructureMember.".format(member))
+
+    @staticmethod
+    def create(member_name, child):
```

Comment:
Rename "child" to "inner_member" or "nested_member" ?

Reply:
Done.